### ISSUE TITLE
Fabric UI doesn't work with Mobx

### ISSUE THREAD

**Issue by mininternet** [OfficeDev/office-ui-fabric-react#930]
### Bug Report

 "office-ui-fabric-react": "^1.3.7",
 "mobx": "^3.0.2",
 "mobx-react": "^4.1.0"

#### Describe the issue:

Trying to use the Panel component.

The Panel appears, as expected, when passed a "true" value to the isOpen prop, from the parent container.

However, once opened, the Panel does not close or respond to any setState commands from within the Panel component, nor does the onDismiss event fire.

No errors thrown, just doesn't respond to any attempts to close, either through setState, or @observable Mobx atttributes.

#### Expected behavior:

Should close when clicking the overlay, doesn't respond to any events, or throw errors.

### If applicable, please provide a codepen repro:

import React, { Component } from 'react'
import { observable } from 'mobx'

import { observer, inject } from 'mobx-react'

import { Panel } from 'office-ui-fabric-react/lib/Panel'

export default class FilterPanel extends Component {
@observable filterPanelOpen
  
  constructor(props) {
    super(props)
    this.filterPanelOpen = this.props.active
  }

	closeFilter() {
		this.filterPanelOpen = false
	}

  render() {
    return (
      <Panel
        isOpen={ this.filterPanelOpen }
        isLightDismiss={ true }
        onDismiss= { this.closeFilter.bind(this) }
        headerText='Light Dismiss Panel'
      >
        <span className='ms-font-m'>Light Dismiss usage is meant for the Contextual Menu on mobile sized breakpoints.</span>
      </Panel>
    )
  }
}

**Comment by dzearing** [comment]
I'm really confused by this bug. You refer to a Drawer component, which doesn't exist in fabric.

Your example refers to a Panel but doesn't show usage.

Can you please add more details on what you're using, and how you're using it?

**Comment by dzearing** [comment]
I'm updating the title to be more specific to an issue your having, how i interpret it... but please clarify.

**Comment by mininternet** [comment]
(Apologies, i used the term Drawer since that's what it's called in both Material UI and React Toolbox. I've updated my initial description.)

Basically, the Panel appears when a parent component passes a "true" value to the Panel isOpen prop, but once opened, the Panel does not respond to any attempts to close, either by using setState = false, or by toggling an @observable true/false attribute from within the Panel component.

Nor does the onDismiss event fire. No errors, just unresponsive behavior, where the Panel remains open.

I'm guessing this library has been tested with Redux, but I believe Mobx is becoming more and more popular, especially for those of us with .NET backgrounds.

**Comment by mininternet** [comment]
@dzearing - You can keep this issue on hold for now, I'm going to attempt a couple of new React state management methods to see if I can get Fabric UI to work.

Thanks

**Comment by dzearing** [comment]
I don't think this is specific to mobx or redux, but perhaps how the state is being passed into the Panel.

Here's a codepen of usage: http://codepen.io/dzearing/pen/qqWXYP

Would it be possible for you to fork to include an example of mobx being used here?

**Comment by dzearing** [comment]
I've updated the codepen to include a `setInterval` to toggle the isOpen boolean on and off just to illustrate it opening/closing.

**Comment by dzearing** [comment]
Or at least... paste a code snipped of the props you're passing into the Panel?

**Comment by mininternet** [comment]
Thanks for the Codepen, is very useful. The documentation on the FabricUI site is quite odd compared to most React libraries (import from .../index, React.Component<any, any>, etc.)

I will try again with your code example.

Thanks

**Comment by mininternet** [comment]
Still not working. I agree probably doesn't relate to Mobx, but rather the Panel component's own state mechanism.

When passing a True value as a prop, from the Parent control to the Panel's isOpen property, the Panel displays, as intended.

Yet within the Panel component, neither this.state.isOpened = this.props.isOpened, nor this.state.isOpened = true, nor @isOpened = this.props.isOpened, neither of those assignments are recognized by the Panel's isOpen property. 

The Panel only displays when the isOpen prop = this.props.isOpened. 

TLDR - Panel only recognizes this.props, not this.state. Nor does the onDismiss event fire.

Including code below, with both approaches.

```import React, { Component } from 'react'
import { Panel } from 'office-ui-fabric-react/lib/Panel'

import { observable } from 'mobx'

export default class FilterDrawer extends Component {
@observable isOpened

  constructor(props) {
    super(props)
    this.state = {isOpened: this.props.isOpened}
    this.isOpened = this.props.isOpened
  }

  togglePanel(){
    this.setState({ isOpened: !this.state.isOpened })
    this.isOpened = !isOpened
  }

  render() {
    return (
      <Panel
        isOpen={this.state.isOpened}
        isLightDismiss={true}
        onDismiss= {(e) => this.togglePanel(this) }
        headerText='Panel - Small, left-aligned, fixed'
      >
        <h1>this is the header</h1>
      </Panel>
    )
  }
}
```

And here is the transpiled ES5 code that i've stepped through - 

```
'use strict';
Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.default = undefined;
var _createClass = function() {
    function defineProperties(target, props) {
        for (var i = 0; i < props.length; i++) {
            var descriptor = props[i];
            descriptor.enumerable = descriptor.enumerable || false;
            descriptor.configurable = true;
            if ("value"in descriptor)
                descriptor.writable = true;
            Object.defineProperty(target, descriptor.key, descriptor);
        }
    }
    return function(Constructor, protoProps, staticProps) {
        if (protoProps)
            defineProperties(Constructor.prototype, protoProps);
        if (staticProps)
            defineProperties(Constructor, staticProps);
        return Constructor;
    }
    ;
}();
var _desc, _value, _class, _descriptor;
var _react = __webpack_require__(81);
var _react2 = _interopRequireDefault(_react);
var _Panel = __webpack_require__(1718);
var _mobx = __webpack_require__(694);
function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _initDefineProp(target, property, descriptor, context) {
    if (!descriptor)
        return;
    Object.defineProperty(target, property, {
        enumerable: descriptor.enumerable,
        configurable: descriptor.configurable,
        writable: descriptor.writable,
        value: descriptor.initializer ? descriptor.initializer.call(context) : void 0
    });
}
function _classCallCheck(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
        throw new TypeError("Cannot call a class as a function");
    }
}
function _possibleConstructorReturn(self, call) {
    if (!self) {
        throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
    }
    return call && (typeof call === "object" || typeof call === "function") ? call : self;
}
function _inherits(subClass, superClass) {
    if (typeof superClass !== "function" && superClass !== null) {
        throw new TypeError("Super expression must either be null or a function, not " + typeof superClass);
    }
    subClass.prototype = Object.create(superClass && superClass.prototype, {
        constructor: {
            value: subClass,
            enumerable: false,
            writable: true,
            configurable: true
        }
    });
    if (superClass)
        Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass;
}
function _applyDecoratedDescriptor(target, property, decorators, descriptor, context) {
    var desc = {};
    Object['ke' + 'ys'](descriptor).forEach(function(key) {
        desc[key] = descriptor[key];
    });
    desc.enumerable = !!desc.enumerable;
    desc.configurable = !!desc.configurable;
    if ('value'in desc || desc.initializer) {
        desc.writable = true;
    }
    desc = decorators.slice().reverse().reduce(function(desc, decorator) {
        return decorator(target, property, desc) || desc;
    }, desc);
    if (context && desc.initializer !== void 0) {
        desc.value = desc.initializer ? desc.initializer.call(context) : void 0;
        desc.initializer = undefined;
    }
    if (desc.initializer === void 0) {
        Object['define' + 'Property'](target, property, desc);
        desc = null;
    }
    return desc;
}
function _initializerWarningHelper(descriptor, context) {
    throw new Error('Decorating class property failed. Please ensure that transform-class-properties is enabled.');
}
var FilterDrawer = (_class = function(_Component) {
    _inherits(FilterDrawer, _Component);
    function FilterDrawer(props) {
        _classCallCheck(this, FilterDrawer);
        var _this = _possibleConstructorReturn(this, (FilterDrawer.__proto__ || Object.getPrototypeOf(FilterDrawer)).call(this, props));
        _initDefineProp(_this, 'isOpened', _descriptor, _this);
        _this.state = {
            isOpened: _this.props.active
        };
        _this.isOpened = _this.props.active;
        return _this;
    }
    _createClass(FilterDrawer, [{
        key: 'togglePanel',
        value: function togglePanel() {
            this.setState({
                isOpened: !this.state.isOpened
            });
            this.isOpened = !isOpened;
        }
    }, {
        key: 'render',
        value: function render() {
            var _this2 = this;
            return _react2.default.createElement(_Panel.Panel, {
                isOpen: this.state.isOpened,
                isLightDismiss: true,
                onDismiss: function onDismiss(e) {
                    return _this2.togglePanel(_this2);
                },
                headerText: 'Panel - Small, left-aligned, fixed'
            }, _react2.default.createElement('h1', null, 'this is the header'));
        }
    }]);
    return FilterDrawer;
}(_react.Component),
(_descriptor = _applyDecoratedDescriptor(_class.prototype, 'isOpened', [_mobx.observable], {
    enumerable: true,
    initializer: null
})),
_class);
exports.default = FilterDrawer;
;var _temp = function() {
    if (typeof __REACT_HOT_LOADER__ === 'undefined') {
        return;
    }
    __REACT_HOT_LOADER__.register(FilterDrawer, 'FilterDrawer', 'C:/Users/a/Documents/Visual Studio 2015/Projects/whogoverns.us/src/components/organisms/FilterDrawer/index.js');
}();
;//////////////////
// WEBPACK FOOTER
// ./src/components/organisms/FilterDrawer/index.js
// module id = 777
// module chunks = 0

```

**Comment by cschleiden** [comment]
Haven't used mobx before, but how do you change state when `this.isOpened` changes? Looks like this is independent right now?

**Comment by mininternet** [comment]
I was including both code samples for reference, to indicate i attempted both.

Meaning i tried isOpen={this.isOpened}, but also didn't work.

With Mobx, you declare a local variable with a decorater, @observable, and any update to that variable triggers a re-render.

So with Material-UI, i'm able to change @isOpened = true, on an event click, and the component recognizes that updated variable..

(Also, using an @inject decorator, you can bind the local component variables to a global store, which IMO is far less boilerplate than Redux actions+reducers+stores, but that isn't the issue here...)

**Comment by cschleiden** [comment]
Where do you add `@observer` to your component? Quickly put together a codepen http://codepen.io/cschleiden2/pen/BpVPBW seems to work just fine.

**Comment by mininternet** [comment]
Looking at your Codepen, i notice it's using TypeScript as the preprocesser. 

When I change it to Babel (i don't really see a lot of TypeScript in the React ecosystem), i get the same behavior i mentioned, where the Panel doesn't display, nor react to any updating of isOpened. Nor does it throw any errors. 

I'm guessing it's an issue with TypeScript / Babel ES6 transpilation (this). 

Here is the code, using the Panel component -

```
const { observable } = mobx;
const { observer } = mobxReact;

@observer 
class Content extends React.Component {
  @observable isOpened = false;

 componentDidMount() {
    setInterval(() => {
      this.isOpened = !this.isOpened
    }, 1500)
  }

  render() {
    return (
  <Fabric.Fabric className='foo'>
        <Fabric.Button 
          onClick={ this.isOpened = !this.isOpened }    
        >{this.isOpened ? 'Collapse panel' : 'Expand panel'}</Fabric.Button>
          <Fabric.Panel
            isOpen={ this.isOpened }
            isLightDismiss={ true }
            headerText='Panel - Small, left-aligned, fixed'
          >
            <Fabric.Button>I'm a focus target</Fabric.Button>
            <span className='ms-font-m'>Content goes here.</span>
          </Fabric.Panel>
       
      </Fabric.Fabric>
    );
  }
}

ReactDOM.render( 
  <Content />,
  document.getElementById('content')
);
```

For reference, here is the boilerplate I'm using, with Babel as the default transpiler. 

https://github.com/diegohaz/arc

**Comment by cschleiden** [comment]
http://codepen.io/cschleiden2/pen/PWBGvL with Babel, (mostly) with your code, works just fine. 
If you take your code and look at the console output, you can see, that you are not passing in a callback to `onClick` for the `Button` which triggers a mobX invariant error. 

I don't think there is any problem with mobX and OfficeFabric here. Please re-open if you find another issue.

**Comment by dzearing** [comment]
Dang. I just made a mobx sample and saw @cschleiden did exactly the same. Thanks Chris.

fwiw same outcome :)
http://codepen.io/dzearing/pen/ygqKLp?editors=0010

**Comment by cschleiden** [comment]
Sorry, I couldn't resist :) Hadn't used mobx before and wanted to give it a try.

**Comment by mininternet** [comment]
I understsnd, Mobx is interesting.
My project still has the same issue, not replicable in that Codepen, but my Webpack config obviously won't be the same. Material UI works as expected.
Material UI is still the gold standard as far as easy integration with React goes, but I'll try to circle back to Fabric UI once it's not so bleeding edge.
Thanks for the assistance. 

-------- Original message --------
